src/Pinch/Protocol/Compact.hs

```diff
@@ -0,0 +1,487 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE CPP                 #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+-- |
+-- Module      :  Pinch.Protocol.Compact
+-- Copyright   :  (c) Ben Gamari 2015
+-- License     :  BSD3
+--
+-- Maintainer  :  Abhinav Gupta <mail@abhinavg.net>
+-- Stability   :  experimental
+--
+-- Implements the Thrift Compact Protocol as a 'Protocol'.
+module Pinch.Protocol.Compact (compactProtocol) where
+
+
+#if __GLASGOW_HASKELL__ < 709
+import Control.Applicative
+#endif
+
+import Control.Monad
+import Data.List           (sortBy)
+import Data.Ord            (comparing)
+import Data.Bits           hiding (shift)
+import Data.ByteString     (ByteString)
+import Data.Foldable       (null)
+import Data.HashMap.Strict (HashMap)
+import Data.Int            (Int16, Int32, Int64)
+import Data.Word           (Word64, Word8)
+import Data.Monoid
+
+import qualified Data.ByteString     as B
+import qualified Data.HashMap.Strict as M
+import qualified Data.Text.Encoding  as TE
+
+import Pinch.Internal.Builder (Builder)
+import Pinch.Internal.Message
+import Pinch.Internal.Parser  (Parser, runParser, runParser')
+import Pinch.Internal.TType
+import Pinch.Internal.Value
+import Pinch.Protocol         (Protocol (..))
+
+import qualified Pinch.Internal.Builder  as BB
+import qualified Pinch.Internal.FoldList as FL
+import qualified Pinch.Internal.Parser   as P
+
+
+-- | Provides an implementation of the Thrift Compact Protocol.
+compactProtocol :: Protocol
+compactProtocol = Protocol
+    { serializeValue     = binarySerialize
+    , deserializeValue'  = binaryDeserialize ttype
+    , serializeMessage   = binarySerializeMessage
+    , deserializeMessage = binaryDeserializeMessage
```

Comment:
Just a nit: These methods were called `binary{Serialize,Deserialize}*` because they were implementing the binary protocol. Here, you would want to call them `compactSerialize`, `compactDeserialize`, etc. or just `serialize`, `deserialize`.